### PR TITLE
New commands: sort(), head(), and tail()

### PR DESCRIPTION
**Pull request by nfischer**
These are initial versions of the `head()` and `sort()` commands. **Not ready to merge**. Feel free to take a look and give feedback on what I have implemented so far.

I haven't written unit tests yet, but I will do so soon.

The reason these are together in one PR is that I wanted to be able to hack out a translation of an adaptation of the famous [Donald Knuth shell problem](http://www.leancrew.com/all-this/2011/12/more-shell-less-egg/):

```Bash
#!/bin/bash
cat input.txt |
tr -cs A-Za-z '\n' |
tr A-Z a-z |
sort |
uniq -c |
sort -rn |
head -n $1
```

Will now (correctly) translate to:

```Javascript
#!/usr/bin/env node
require('shelljs/global');
config.silent = true;
echo(cat('input.txt')
  .exec('tr -cs A-Za-z \'\\n\'')
  .exec('tr A-Z a-z')
  .sort() // sort works!
  .exec('uniq -c')
  .sort('-rn') // it even takes options!
  .head('-n', process.argv[2])); // and head() should take options of the style `head({'-n', 5}, 'file.txt')`
```

Also, if someone wants `tail()`, chime in and I can write that out as well.

### CODE REVIEW

**Comment by nfischer** [comment]
I just added in the test cases, so this is ready for review. If anyone from the community wants to chime in for features, please do.

The semantics are slightly different than unix's `sort`: capitalized words are sorted before non-capitalized words in this function, but unix `sort` doesn't prefer them. So we sort things like:

```
Foo
bar
```

While unix sorts things like:

```
bar
Foo
```

This difference could be avoided, but I'm not sure if it really matters so much.

**Comment by nfischer** [comment]
@ariporad If you get the chance, this should be ready for review. I see that it's failing on `v0.10`, but I think this will pass on the other versions. I can investigate `v0.10` if we think it's a priority.

Feel free to review but not merge, since there's no huge rush to get these features. We can always use more input from the community. On the other hand, if we merge now, we can always go back and add features before the next release.

File: README.md [hunk 1]
```diff
@@ -330,6 +330,20 @@ containing the files if more than one file is given (a new line character is
 introduced between each file). Wildcard `*` accepted.
 
 
+### head([{'-n', <num>},] file [, file ...])
+### head([{'-n', <num>},] file_array)
+
+Examples:
+
+```javascript
+var str = head({'-n, 1}, 'file*.txt');
```
Comment (ariporad) [inline README.md]: You're missing a quote her.

File: README.md [hunk 2]
```diff
@@ -330,6 +330,20 @@ containing the files if more than one file is given (a new line character is
 introduced between each file). Wildcard `*` accepted.
 
 
+### head([{'-n', <num>},] file [, file ...])
+### head([{'-n', <num>},] file_array)
```
Comment (ariporad) [inline README.md]: Sorry if this is stupid, but what does head do? And maybe we should include a description here.

File: README.md (hunk 1 above)
Comment (nfischer) [inline README.md]: I'll fix

File: src/sort.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  }
+  if (options.numerical) {
+    sorted = lines.sort(function(a, b) {
+      var objA = parseNumber(a);
+      var objB = parseNumber(b);
+      if (objA.hasOwnProperty('num') && objB.hasOwnProperty('num')) {
+        return (objA.num !== objB.num) ? (objA.num - objB.num)
+            : (objA.value.localeCompare(objB.value));
+      } else {
+        return (objA.value.localeCompare(objB.value));
```
Comment (ariporad) [inline src/sort.js]: What's localeCompare?

File: README.md (hunk 2 above)
Comment (nfischer) [inline README.md]: I'll update the description (I just realized I forgot to do that...)

**Comment by nfischer** [comment]
Looks like there's iojs compatibility issues. I'll try to fix that today and see if I can get `v0.10` working as well. I may also try to get unix semantics (I think I know how to do that).

**Comment by ariporad** [comment]
A few nits, otherwise looks good.

File: src/sort.js (hunk 3 above)
Comment (nfischer) [inline src/sort.js]: It's the default string comparison. It returns `-1` if `this` is "less than" (sorted before) the argument.

So:

```
'foo'.localeCompare('bar'); // 1
'foo'.localeCompare('zed'); // -1
'foo'.localeCompare('foo'); // 0
```

File: README.md (hunk 1 above)
Comment (nfischer) [inline README.md]: Should be fixed now (good catch)

File: README.md (hunk 2 above)
Comment (nfischer) [inline README.md]: Check the description now

**Comment by nfischer** [comment]
Ok, I think I fixed the inconsistency. Also, this is now compatible with how unix sort works, which is a plus. Should be good to merge now.

**Comment by nfischer** [comment]
Let's hold off to merge this. I think this should be refactored to read input in via buffers (vs. reading it all at once), when possible. This is to account for the case where a file is infinite in size (or maybe just super large) and we print out the first 10 lines, keeping the operation constant time instead of hanging indefinitely.

I should have time to refactor this today. I can also add in `tail()` to the mix in another PR, for the sake of completeness (`tail` prints the last 10 lines, as opposed to `head` which prints the first 10 lines).

**Comment by ariporad** [comment]
SGTM!

**Comment by nfischer** [comment]
@ariporad I refactored this into 3 commits. `tail()` is a new command. This also modifies `head()` so that it only reads as much of the file as necessary (so it works for files that are infinite in length, like if you run `yes > file.txt`).

Leave any comments/concerns and I can address when I get back.

**Comment by nfischer** [comment]
@ariporad This should be ready for review

File: src/sort.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        (objA.num - objB.num) :
+        unixCmp(objA.value, objB.value));
+  } else {
+    return unixCmp(objA.value, objB.value);
+  }
+}
+
+//@
+//@ ### sort([options,] file [, file ...])
+//@ ### sed([options,] file_array)
```
Comment (ariporad) [inline src/sort.js]: Why is this also called `sed`? I thought `sed` was find/replace?

**Comment by ariporad** [comment]
@nfischer: One comment, also can you either squash your final commit or make it comply with the commit guidelines. Thanks!

File: src/sort.js (hunk 4 above)
Comment (nfischer) [inline src/sort.js]: Whoops! That was a mistake. I'll fix that

**Comment by nfischer** [comment]
@ariporad Rebased this and fixed the issues

**Comment by nfischer** [comment]
@ariporad Let me know once you've taken a look at this. I'll address any issues and rebase. If there are no issues, feel free to give the thumbs up and I'll rebase and merge.

**Comment by ariporad** [comment]
LGTM!

**Comment by nfischer** [comment]
Ok great! I'll rebase this and merge

**Comment by nfischer** [comment]
LGTM (because of rebase)

**Comment by facundoolano** [comment]
any chance of having this published to npm?

**Comment by nfischer** [comment]
Hopefully soon. We had some last minute concerns raised, but it should be resolved soon

**Pull request merged**